### PR TITLE
duration till next ticket price

### LINKED ISSUE #480: duration till next ticket price
add duration till next ticket price as show in the screenshot below
Screenshot 2021-06-18 at 16 13 50

`See All` in the current implementation should be `See all`. Note the `all` is in all lower case
Screenshot 2021-06-18 at 16 15 00

### PR DESCRIPTION
**Pull request by song50119**
Resolves #480

### CODE REVIEW

File: ui/load/load.go [hunk 1]
```diff
@@ -55,7 +55,7 @@ type Icons struct {
 	AboutIcon, DebugIcon, VerifyMessageIcon, LocationPinIcon, AlertGray, ArrowDownIcon,
 	WatchOnlyWalletIcon, CurrencySwapIcon, SyncingIcon, ProposalIconActive, ProposalIconInactive,
 	Restore, DocumentationIcon, DownloadIcon, TimerIcon, TicketIcon, TicketIconInactive, StakeyIcon,
-	List, ListGridIcon, DecredSymbolIcon *widget.Image
+	List, ListGridIcon, DecredSymbolIcon, TimeIcon *widget.Image
```
Comment (Sirmorrison) [inline ui/load/load.go]: Kindly remove the `TimeIcon` we already have a `TimerIcon`

File: ui/load/load.go [hunk 2]
```diff
@@ -169,6 +169,7 @@ func NewLoad(th *decredmaterial.Theme, decredIcons map[string]image.Image) *Load
 		List:                       &widget.Image{Src: paint.NewImageOp(decredIcons["list"])},
 		ListGridIcon:               &widget.Image{Src: paint.NewImageOp(decredIcons["list_grid"])},
 		DecredSymbolIcon:           &widget.Image{Src: paint.NewImageOp(decredIcons["decred_symbol"])},
+		TimeIcon:                   &widget.Image{Src: paint.NewImageOp(decredIcons["time"])},
```
Comment (Sirmorrison) [inline ui/load/load.go]: Kindly remove the TimeIcon we already have a TimerIcon

File: ui/page/tickets/utils.go [hunk 3]
```diff
@@ -386,3 +387,34 @@ func createClickGestures(count int) []*gesture.Click {
...
+	if allsecs > 59 {
+		mins := allsecs / 60
+		allsecs %= 60
+		str += fmt.Sprintf("%dm ", mins)
+	}
+	if allsecs > 0 {
+		str += fmt.Sprintf("%ds ", allsecs)
+	}
+	return str
+}
```
Comment (Sirmorrison) [inline ui/page/tickets/utils.go:494]:
```suggestion
	var str string
	if allsecs > 604799 {
		weeks := allsecs / 604800
		allsecs %= 604800
		str = fmt.Sprintf("%dw ", weeks)
	}
	if allsecs > 86399 {
		days := allsecs / 86400
		allsecs %= 86400
		str = fmt.Sprintf("%dd ", days)
	}
	if allsecs > 3599 {
		hours := allsecs / 3600
		allsecs %= 3600
		str = fmt.Sprintf("%dh ", hours)
	}
	if allsecs > 59 {
		mins := allsecs / 60
		allsecs %= 60
		str = fmt.Sprintf("%dm ", mins)
	}
	if allsecs > 0 {
		str = fmt.Sprintf("%ds ", allsecs)
	}
	return str
}
```
I think using a switch case would be better.

Also, there is a TimeAgo() in the component package... kindly make use of that. 
https://github.com/planetdecred/godcr/blob/269088483ecb35ae68550985dd6b39ae4df25591/ui/page/components/components.go#L280

File: wallet/commands.go [hunk 4]
```diff
@@ -284,8 +284,8 @@ func (wal *Wallet) GetTransaction(walletID int, txnHash string) {
 			wal.Send <- resp
 			return
 		}
-
-		txn, err := wall.GetTransactionRaw(hash[:])
+		txn, err := wall.GetTransactionRaw(hash.String())
+		log.Info(hash.String(), txn.Amount)
```
Comment (Sirmorrison) [inline wallet/commands.go]: This change here is not part of the task are working on is it?

File: ui/page/tickets/overview.go [hunk 5]
```diff
@@ -125,9 +125,32 @@ func (pg *Page) ticketPriceSection(gtx layout.Context) layout.Dimensions {
...
 					Bottom: values.MarginPadding11,
 				}.Layout(gtx, func(gtx C) D {
-					tit := pg.Theme.Label(values.TextSize14, "Ticket Price")
-					tit.Color = pg.Theme.Color.Gray2
-					return pg.titleRow(gtx, tit.Layout, material.Switch(pg.Theme.Base, pg.autoPurchaseEnabled).Layout)
+					leftWg := func(gtx C) D {
+						return layout.Flex{Alignment: layout.Middle}.Layout(gtx,
+							layout.Rigid(func(gtx C) D {
+								tit := pg.Theme.Label(values.TextSize14, "Ticket Price")
+								tit.Color = pg.Theme.Color.Gray2
```
Comment (Sirmorrison) [inline ui/page/tickets/overview.go]: A better variable name will be `title` or `text` but not `tit`.

File: ui/page/tickets/overview.go [hunk 6]
```diff
@@ -125,9 +125,32 @@ func (pg *Page) ticketPriceSection(gtx layout.Context) layout.Dimensions {
...
+							}),
+							layout.Rigid(func(gtx C) D {
+								secs, _ := pg.WL.MultiWallet.NextTicketPriceRemaining()
+								txt := pg.Theme.Label(values.TextSize14, nextTicketRemaining(int(secs)))
+								txt.Color = pg.Theme.Color.Gray2
+								return txt.Layout(gtx)
+							}),
+						)
+					}
+					return pg.titleRow(gtx, leftWg, material.Switch(pg.Theme.Base, pg.autoPurchaseEnabled).Layout)
```
Comment (Sirmorrison) [inline ui/page/tickets/overview.go]: use decredmaterial Switch as it is customized to the project. You can see example usage on the settings page.

**Review by Sirmorrison — CHANGES_REQUESTED** [review]

File: wallet/commands.go (hunk 4 above)
Comment (song50119) [reply wallet/commands.go]: This PR have to update to latest version of dcrlibwallet, some function have to change to compatible with upstream

File: ui/page/tickets/utils.go (hunk 3 above)
Comment (song50119) [reply ui/page/tickets/utils.go:494]: TimeAgo is calculate base on timestamp, so it won't work like this

Comment (song50119) [reply ui/page/tickets/utils.go:494]: [this is how on dcrdata calculate](https://github.com/decred/dcrdata/blob/091b91b235e896a06ab086251892c1f6b6d7a366/explorer/templates.go#L200-L228)

File: ui/page/tickets/overview.go [hunk 7]
```diff
@@ -52,7 +51,7 @@ func NewTicketPage(l *load.Load) *Page {
 		ticketPageContainer: &layout.List{Axis: layout.Vertical},
 		purchaseTicket:      l.Theme.Button(new(widget.Clickable), "Purchase"),
 
-		autoPurchaseEnabled: new(widget.Bool),
+		autoPurchaseEnabled: l.Theme.Switch(),
```
Comment (Sirmorrison) [inline ui/page/tickets/overview.go:54]: the `autoPurchaseEnabled` feature seem not to have been implemented. Kindly disable the button toggle.
`pg.autoPurchaseEnabled.Disabled()` in the `OnResume()` method

**Review by Sirmorrison — CHANGES_REQUESTED** [review]

File: ui/page/tickets/overview.go (hunk 7 above)
Comment (song50119) [reply ui/page/tickets/overview.go:54]: Would be great if create new PR for that function

Comment (Sirmorrison) [reply ui/page/tickets/overview.go:54]: Switch widget disabled() already exist
https://github.com/planetdecred/godcr/blob/1dea6510eb61864cdfc28fe4a74973411a61a78e/ui/page/privacy_page.go#L56

Comment (Sirmorrison) [reply ui/page/tickets/overview.go:54]: > the `autoPurchaseEnabled` feature seem not to have been implemented. Kindly disable the button toggle.
> `pg.autoPurchaseEnabled.Disabled()` in the `OnResume()` method

image

What was requested was to have the auto purchase toggle disabled not removed.
Kindly add it back as it is a vital part of the UI... all you need to so is the following.

```
func (pg *Page) OnResume() {
	pg.ticketPrice = dcrutil.Amount(pg.WL.TicketPrice()).String()
	go pg.WL.GetVSPList()
         // TODO: automatic ticket purchase functionality
        pg.autoPurchaseEnabled.Disabled()
}
```

**Review by Sirmorrison — APPROVED** [review]

**Review by oshorefueled — APPROVED** [review]

**Pull request merged**